Create an import section in settings and add importing of bikes.csv data from a Strava bulk export

F-Stop opened the pull request:
This pull request is intended to resolve issue #262 (adding a dedicated "import" section to the Settings menu) and issue #261 (create the ability to import bike data from a Strava bulk export).  These changes also set the stage for addressing issue #242, since importing bikes and having a dedicated import section are precursors to a good implementation of importing Strava activity metadata with Strava bulk-export activities.

Adding a dedicated import section to the Settings menu required creating new user interface text, so this likely will need some translation work.   

If someone can get me a shoes.csv file to play with, I should be able to easily also create a shoe importing function.

F-Stop commented on backend/app/gears/gear/router.py:
```diff
@@ -263,3 +271,99 @@ async def delete_gear(
+
+        # CSV file location
+        bulk_import_dir = core_config.FILES_BULK_IMPORT_DIR
+        bikesfilename = "bikes.csv" # Hard coding filename for now (this is the filename Strava uses)
```
Want to move this to core/config.py?  I will have a few more filenames as the importing work moves forward (shoes.csv, activities.csv, etc.), so having them all in one place would likely be handy.

F-Stop commented:
Alright @joaovitoriasilva, think I have got this ready to go for ya.  

The import section will get a lot more fun once we pull in the full activities import code (which I am still working on [here](https://github.com/F-Stop/endurain/tree/strava-bulk-import-v13)).

F-Stop commented:
Okay, I have much better duplicate checking logic now in here now, based on (but improved from) the shoe importer logic in #318.  Created a small function in gears.util to check for duplicates, and also am now checking for overlapping nicknames in a single import (which was also causing a crash - say if someone had "Susi" and "susi").  Will move this same logic over to the shoe import in #318 in a bit.  Log entries are also more informative for the user, letting them know of multiple potential reasons why the gear was not imported (nickname alone overlapped, entire gear item overlapped, and entire gear item overlapped because it was already present due to a Strava link).

joaovitoriasilva commented:
@F-Stop some notes, I started reviewing this and:
 - New route will be moved from gears to Strava router, because the imported job is related to Strava and I will rename it from `/stravabikesimport` to probably `/import/bikes`;
 - I moved the CSV parsing to a separate function in strava/gear_utils.py for better readability and separation. I believe this will also be better in the future when a full Strava import is added to call the specific modules when needed, reducing repeated code;
 - I see that you check if there is a gear with that nickname inserted for that user for each entry in the CSV, however the gear create logic already checks this (`create_gear` in `gears crud.py`) so I am finding this redundant and I will probably remove it. Any comments on this?

F-Stop commented:
>     * New route will be moved from gears to Strava router, because the imported job is related to Strava and I will rename it from `/stravabikesimport` to probably `/import/bikes`;

Sure!  Do the same thing for shoes in #318 ;)

>     * I moved the CSV parsing to a separate function in strava/gear_utils.py for better readability and separation. I believe this will also be better in the future when a full Strava import is added to call the specific modules when needed, reducing repeated code;

Sure, why not?  To allow it to be used for both bikes and shoes, the function should get passed the list of items to check for in the header row, as that is different between bikes ('Bike Name', 'Bike Brand', and 'Bike Model') and shoes ('Shoe Name', 'Shoe Brand', 'Shoe Model'); see #318 for the matching shoe importer code.

>     * I see that you check if there is a gear with that nickname inserted for that user for each entry in the CSV, however the gear create logic already checks this (`create_gear` in `gears crud.py`) so I am finding this redundant and I will probably remove it. Any comments on this?

If I recall right, I ran into a crash when duplicates were not checked for before passing things off to create_gear. I can't recall right now, but I think it especially happened when the item was duplicate in all ways except case (e.g., "Ox" and "ox"; I think create_gear doesn't deal with that case), but it may also have happened due to challenges with raising an exception in a background process (that we ran into before).  In any case, I decided to just check for duplicates in the import function, which also nicely added the ability to have the logs (and some day, notifications) report much more complete information on what the duplicate is and why it was getting rejected (duplicate nickname, totally duplicate gear, gear already present due to a Strava link, etc.).

Fine with whatever you want :)

joaovitoriasilva commented:
Hi @F-Stop just pushed a commit with some updates inline with my previous comment and your notes. With the review of shoes import logic, some code will need to be adjusted, but one problem at a time. Would love your feedback on the changes. 

Thank you

joaovitoriasilva reviewed (approved):
Tested again and seems to be inline with what is intended. Will merge with v0.15.0

F-Stop commented:
Finally got a chance to look this over. @joaovitoriasilva   

1 - Now that this code removes the check for case-insensitive naming of gear, we should probably update get_gear_user_by_nickname and get_gear_user_contains_nickname to reflect the new schema.  They both still use:

`
unquote(nickname).replace("+", " ").lower()
`

This does not affect the import, but it does allow some odd behavior in the GUI.  For example, you can now get items with duplicate nicknames.  

Test procedure:

- First add a gear with the nickname "NameWithPlus+" in the gui.
- That adds a gear named "NameWithPlus", as expected (replace + with " " and then strip)
- Trying to add a new gear item with nickname "NameWithPlus" in the gui results in a correct "duplicate name" message.  
- But trying to add a near gear item with nickname "NameWithPlus+" allows addition of the new gear, and adds a second gear item named "NameWithPlus"

<img width="372" height="229" alt="image" src="https://github.com/user-attachments/assets/986151b2-efca-4de9-999e-25481485faee" />

Fix:

So, we should probably at least add the .strip() to the checks.

2 - The name checking routine in create_multiple_gears was not correctly filtering out invalid names.  So, the following were actually importing:
`
,test5,should NOT import
 ,test6, should NOT import
++, test7, should NOT import
`

I have pushed out changes that should fix these two things :)

F-Stop commented:
Keeping the lowercase bits in get_gear_user_by_nickname resulted in improper importing of duplicate names, so I have removed the lowercasing of items in get_gear_user_by_nickname.  Also updated the docs a bit.

joaovitoriasilva commented on backend/app/gears/gear/crud.py:
- Changes to code on line 116 make sense;
- Changes on line 156 in my view does not make sense. I will apply the same logic on 116 and compare the two lower strings like on function `get_gear_user_contains_nickname`;
- Line 304 and 305 will not apply and continue to use the previous code.

joaovitoriasilva reviewed (approved):
Approved but keep in mind comments on gears crud

F-Stop commented:
see #361
